src/components/application_manager/src/policies/policy_handler.cc

```diff
@@ -1501,12 +1501,31 @@ void PolicyHandler::OnSnapshotCreated(const BinaryMessage& pt_string) {
     return;
   }
 
-  AppIdURL app_url = policy_manager_->GetNextUpdateUrl(urls);
-  while (!IsUrlAppIdValid(app_url.first, urls)) {
-    app_url = policy_manager_->GetNextUpdateUrl(urls);
+  const uint32_t app_id = GetAppIdForSending();
+  const std::string app_policy_id =
+      application_manager_.application(app_id)->policy_app_id();
+
+  const auto is_appid_or_default = [app_policy_id](const EndpointData& ed) {
+    return helpers::Compare<std::string, helpers::EQ, helpers::ONE>(
+        ed.app_policy_id, policy::kDefaultId, app_policy_id);
+  };
+
+  EndpointUrls::const_iterator app_item =
+      std::find_if(urls.begin(), urls.end(), is_appid_or_default);
+
+  if (app_item == urls.end() || app_item->url.empty()) {
+    LOG4CXX_ERROR(logger_, "Application URLs unavailable!");
+    return;
   }
-  const std::string& url = urls[app_url.first].url[app_url.second];
-  SendMessageToSDK(pt_string, url);
+
+   // Set the ulr index to zero if this is the last ulr in the app queue
+  if (retry_sequence_[app_id] >= app_item->url.size()) {
+    retry_sequence_[app_id] = 0;
+  }
+
+  uint32_t& retry_url_id = retry_sequence_[app_id];
```

Comment:
@ZhdanovP Why do we need reference here?

Reply:
@ZhdanovP Please explain in details what is convenience in this case?

Reply:
@AByzhynar @EKuliiev for increment it in the next line 
```const std::string url_to_send = app_item->url[retry_url_id++];```

Reply:
@ZhdanovP 
Maybe better for this case when retry_url_id is referenced the only one time to prefer next construction 
const std::string url_to_send = app_item->url[retry_sequence_[app_id]++];